great_expectations/core/datasource_dict.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+
+import copy
+import logging
+from collections import UserDict
+from typing import TYPE_CHECKING
+
+import great_expectations.exceptions as gx_exceptions
+from great_expectations.data_context.types.base import (
+    datasourceConfigSchema,
+)
+from great_expectations.datasource.new_datasource import BaseDatasource
+
+if TYPE_CHECKING:
+    from great_expectations.core.config_provider import _ConfigurationProvider
+    from great_expectations.data_context.data_context.abstract_data_context import (
+        AbstractDataContext,
+    )
+    from great_expectations.data_context.store.datasource_store import DatasourceStore
+    from great_expectations.datasource.fluent import Datasource as FluentDatasource
+
+logger = logging.getLogger(__name__)
+
+
+class DatasourceDict(UserDict):
+    """
+    An abstraction around the DatasourceStore to enable easy retrieval and storage of Datasource objects
+    using dictionary syntactic sugar.
+    """
+
+    def __init__(
+        self,
+        context: AbstractDataContext,
+        datasource_store: DatasourceStore,
+        config_provider: _ConfigurationProvider,
+    ):
+        self._context = context  # If possible, we should avoid passing the context through - once block-style is removed, we can extract this
```

Comment:
This is only used for block-style - we can remove and have proper dependency injection once we remove support for these legacy datasources